pulp_rpm/app/models.py

```diff
@@ -1,33 +1,203 @@
+    changelogs = models.TextField(default='[]', blank=True)
 
-    # field1 = models.TextField(blank=False, null=False)
-    # field2 = models.TextField(blank=False, null=False)
+    # A string containing a JSON-encoded list of dictionaries, each of which represents a single
+    # file. Each file dict contains the following fields:
+    #
+    #   type (str):     one of "" (regular file), "dir", "ghost"
+    #   path (str):     path to file
+    #   name (str):     filename
+    files = models.TextField(default='[]', blank=True)
 
-    # class Meta:
-    #     unique_together = (
-    #         'field1',
-    #         'field2'
-    #     )
+    # Each of these is a string containing a JSON-encoded list of dictionaries, each of which
+    # represents a dependency. Each dependency dict contains the following fields:
+    #
+    #   name (str):     name
+    #   flags (str):    flags
+    #   epoch (str):    epoch
+    #   version (str):  version
+    #   release (str):  release
+    #   pre (bool):     preinstall
+    requires = models.TextField(default='[]', blank=True)
+    provides = models.TextField(default='[]', blank=True)
+    conflicts = models.TextField(default='[]', blank=True)
+    obsoletes = models.TextField(default='[]', blank=True)
+    suggests = models.TextField(default='[]', blank=True)
+    enhances = models.TextField(default='[]', blank=True)
+    recommends = models.TextField(default='[]', blank=True)
+    supplements = models.TextField(default='[]', blank=True)
+
+    location_base = models.TextField(blank=True)
+    location_href = models.TextField(blank=True)
+
+    rpm_buildhost = models.TextField(blank=True)
+    rpm_group = models.TextField(blank=True)
+    rpm_license = models.TextField(blank=True)
+    rpm_packager = models.TextField(blank=True)
+    rpm_sourcerpm = models.TextField(blank=True)
+    rpm_vendor = models.TextField(blank=True)
+    rpm_header_start = models.BigIntegerField(null=True, blank=True)
+    rpm_header_end = models.BigIntegerField(null=True, blank=True)
+
+    size_archive = models.BigIntegerField(null=True, blank=True)
+    size_installed = models.BigIntegerField(null=True, blank=True)
+    size_package = models.BigIntegerField(null=True, blank=True)
+
+    time_build = models.BigIntegerField(null=True, blank=True)
+    time_file = models.BigIntegerField(null=True, blank=True)
+
+    @property
+    def nevra(self):
+        """ Package NEVRA string (Name-Epoch-Version-Release-Architecture)
+        """
+        return "{n}-{e}:{v}-{r}.{a}".format(
+            n=self.name, e=self.epoch, v=self.version, r=self.release, a=self.arch)
+
+    @property
+    def nvra(self):
+        """ Package NVRA string (Name-Version-Release-Architecture)
+        """
+        return "{n}-{v}-{r}.{a}".format(
+            n=self.name, v=self.version, r=self.release, a=self.arch)
+
+    class Meta:
+        unique_together = (
+            'name', 'epoch', 'version', 'release', 'arch', 'checksum_type', 'pkgId'
```

Comment:
Can someone remind me, why we use checksum_type and checksum(pkgId) as a part of the unique index?

Why can't we leave just checksum_type? The only reason to support two packages with the same nevra and checksum_type but different checksum(pkgId) is if one package is wrong and another one is a correct one which was fixed/updated, thus checksum(pkgId) is different. What do you all think? Is it a valid use case?

Why can't we just leave checksum(pkgId) in the unique index? It will cover the use case ^

Reply:
The checksum value gives an easy way to know if the same package has been signed by different gpg keys (or no key at all). That's the primary reason checksum is included in the unit key. It could be better to use information about the signature directly in the unit key, but that's a lot harder to access since it isn't included in any of the XML metadata. You need to inspect the RPM file directly to get at it.

Assuming you keep the checksum value in the unit key, the checksum type isn't very helpful for establishing uniqueness. But it's such a fundamental part of the checksum value, that it would be weird to have a value without knowing the type. So that's why both have been included in the unit key.